### ISSUE TITLE
[RFC]: Sampling path optimization for NPU model runner v1

### ISSUE THREAD

**Issue by freyfwt** [vllm-project/vllm-ascend#9269]
### Motivation.

The current vLLM Ascend `model_runner_v1` sampling path still relies on the older
`vllm.v1.sample` API and `SamplingMetadata`-based flow. Upstream vLLM has since
refactored the GPU sampling path under `vllm/v1/worker/gpu/sample` with
persistent per-request sampling states, a cleaner `SamplerOutput`, and device-side
implementations for common sampling steps. vLLM Ascend already carries NPU
patches and implementations for parts of this newer path in the v2 runner, but
the v2 model runner is not yet a drop-in replacement for all v1 workloads.

Sampling is on the critical decode path. For common production traffic, the
model returns full logits on every rank in a data-parallel group, while sampling
does mostly redundant work on each rank. In many deployments, users also do not
enable optional logits processors such as penalties, bad words, logit bias, or
top-k/top-p/min-p filtering. We can reduce end-to-end decode latency by combining
the newer upstream sampler structure with Ascend-specific sampling optimizations
inside `model_runner_v1`.

This RFC proposes three related changes:

1. Port the refactored upstream GPU sampler architecture into the Ascend v1
   runner through an adapter layer.
2. Add optional batch-parallel sampling inside each existing DP group by slicing
   the logits rows locally on each rank and gathering only the compact sampling
   results.
3. Add a configurable logits-processing mode so users can choose the default
   behavior, a warning-only skip mode for high-throughput workloads, or a future
   fused Triton-Ascend implementation.

The default behavior should remain unchanged.

### Goals

- Reuse the newer upstream vLLM sampling architecture in the Ascend v1 model
  runner without requiring users to switch to the Ascend v2 runner.
- Reduce redundant sampling work across ranks in the same DP group when every
  rank already has full logits.
- Avoid environment variables. All user-facing controls should be provided
  through `additional_config`.
- Keep the default mode behavior-compatible with existing vLLM Ascend sampling.
- Make `logprobs` and speculative decoding part of the required development
  plan. They may land after the first adapter/batch-parallel PRs, but the feature
  should not be considered complete without them.

### Non-goals

- Do not split logits by vocabulary dimension.
- Do not distribute sampling across different DP groups.
- Do not add a `batch_parallel_group` configuration key. The implementation
  should infer the existing DP group.
- Do not add new environment variables for this feature.
- Do not silently fall back from `skip` mode to default logits processing.

### Configuration

Add a nested sampling configuration under `additional_config`:

```python
LLM(
    model=...,
    additional_config={
        "sampling_config": {
            "enable_batch_parallel": True,
            "logits_processing_mode": "default",
        },
    },
)
```

`sampling_config.enable_batch_parallel`

- Type: `bool`
- Default: `False`
- When enabled, sampling work is partitioned across ranks in the current DP
  group.

`sampling_config.logits_processing_mode`

- Type: `str`
- Allowed values: `"default"`, `"skip"`, `"fused"`
- Default: `"default"`

Mode semantics:

- `"default"`: preserve existing sampling semantics using the refactored sampler
  path and Ascend kernels where available.
- `"skip"`: skip pre-sampling logits processing. If a request uses parameters
  that would normally require skipped logits processing, log a warning and
  continue in skip mode. This mode is intentionally a user-selected performance
  mode and should not fall back to default processing.
- `"fused"`: apply logits processing through a future fused Triton-Ascend kernel
  that preserves default semantics while reducing kernel launch overhead and
  intermediate memory traffic.

The skip-compatible fast path should be documented clearly. Parameters that
should trigger a warning include penalties, bad words, logit bias,
allowed-token/min-token masks, custom logits processors, and non-default
filtering such as top-k/top-p/min-p when those operations are bypassed by the
selected implementation.

#### 1. V1-to-v2 sampler adapter

Introduce an adapter layer in the Ascend v1 runner that lets
`model_runner_v1.sample_tokens()` call the refactored upstream sampler API while
preserving v1 runner lifecycle and bookkeeping.

The adapter should bridge current v1 data structures to the fields expected by
the newer sampler:

- request index mapping
- expanded logits mapping for speculative decoding
- cumulative logits counts for logprobs
- positions and input ids for logits processing
- per-request sampling states and staged updates
- output structures compatible with existing v1 bookkeeping

This keeps the main integration point close to the existing `_sample()` method in
`vllm_ascend/worker/model_runner_v1.py`, where logits are already computed before
calling the sampler.

The implementation should reuse existing Ascend v2 sampling kernels and patches
where possible, including the NPU implementations under
`vllm_ascend/worker/v2/sample` and related patching hooks. New compatibility code
should be kept in vLLM Ascend rather than modifying upstream vLLM internals
unless a small upstream change is clearly reusable by other hardware plugins.

#### 2. Batch-parallel sampling inside the DP group

When `enable_batch_parallel` is enabled, each rank in the same DP group should
slice the full logits tensor along the batch/logits-row dimension and run sampling
only for its local slice.

Important invariants:

- Each rank already owns the full logits tensor for its DP group.
- No communication is needed before sampling. Each rank can slice locally.
- Sampling parallelism is scoped to the current DP group only.
- After local sampling, ranks all-gather compact sampling outputs and restore the
  original batch order before existing v1 bookkeeping runs.

A simple deterministic partition can use contiguous row ranges:

```text
start = floor(num_logits * dp_rank / dp_world_size)
end = floor(num_logits * (dp_rank + 1) / dp_world_size)
```

Only compact results should be gathered, such as sampled token ids, `num_sampled`,
`num_nans`, and, in later phases, logprobs tensors and speculative decoding
metadata. Full logits should not be communicated.

For async scheduling and pipeline-parallel cases, the gather should happen before
the existing bookkeeping path consumes `SamplerOutput`, so downstream code sees a
full-batch output matching today's shape contracts.

#### 3. Logits processing modes

The refactored sampler already centralizes logits processing into a sequence of
state-driven operations:

- logit bias and allowed-token masks
- penalties
- bad words
- temperature
- min-p
- top-k/top-p

The proposed `logits_processing_mode` should select the implementation for this
stage:

- `default`: call the equivalent refactored implementation.
- `skip`: bypass this stage and issue a warning when active requests ask for
  skipped features.
- `fused`: call a Triton-Ascend fused logits-processing kernel.

The fused kernel should be introduced after the adapter and batch-parallel shape
contracts are stable. It should preserve default-mode semantics, support mixed
request parameters in a batch, and be validated against the default path with
tolerance checks where floating-point order changes are expected.

### Development plan

Phase 1: refactored sampler adapter

- Add `sampling_config` parsing in `AscendConfig`.
- Add an Ascend v1 adapter for the refactored sampler states and output.
- Wire `model_runner_v1._sample()` to the adapter behind default-disabled config.
- Keep default behavior compatible with current sampling.

Phase 2: batch-parallel sampling for the base path

- Add local logits-row slicing inside the existing DP group.
- Gather sampled token ids, `num_sampled`, and `num_nans`.
- Restore original row order before v1 bookkeeping.
- Validate greedy and stochastic sampling without logprobs/speculative decoding.

Phase 3: logprobs support

- Support `raw_logprobs` and `processed_logprobs` contracts.
- Gather and restore top-k logprob tensors and token ids.
- Cover prompt/decode logprobs and async scheduling paths.

Phase 4: speculative decoding support

- Extend batch-parallel slicing to expanded logits rows used by speculative
  decoding.
- Gather accepted/rejected token outputs and preserve existing draft-token update
  semantics.
- Validate EAGLE, draft-model, and ngram paths that are supported by Ascend v1.

Phase 5: fused logits processing

- Implement the Triton-Ascend fused logits-processing kernel.
- Cover the default processing features listed above.
- Add fallback-free validation for `fused` mode availability and shape support.

### Testing and validation

Unit tests:

- `sampling_config` parsing, defaults, and invalid mode errors.
- Deterministic row partitioning and full-batch restoration for uneven batch
  sizes.
- Warning behavior for `skip` mode when skipped parameters are present.
- Sampler adapter output shape compatibility.

E2E tests:

- Single-card behavior with default mode matches current v1 sampling.
- Multi-rank DP behavior with batch-parallel sampling matches default sampling
  for greedy requests.
- Stochastic sampling smoke tests with fixed seeds.
- Logprobs correctness once Phase 3 lands.
- Speculative decoding correctness once Phase 4 lands.

Performance validation:

- Measure sampling latency and full decode latency separately.
- Compare default, skip, and fused modes.
- Compare `enable_batch_parallel=False` and `True` at different DP sizes and
  batch sizes.
- Report all-gather overhead separately from local sampling time.
- Test on representative Ascend 910B/910C systems with common models and
  production-like decode-heavy workloads.

### Compatibility

- Default configuration preserves existing user-visible behavior.
- `skip` mode is behavior-changing by design and should require explicit user
  opt-in.
- `fused` mode should be behavior-compatible with default mode.
- No public API change is required outside `additional_config`.
- The feature should continue to work with the latest vLLM main-compatible
  vLLM Ascend branch policy.

### Alternatives considered

Vocabulary-dimension logits sharding:

This was rejected. The intended optimization is not tensor-parallel vocabulary
sharding. Each DP-rank already has full logits, and the optimization should split
work by logits rows inside the DP group.

Pre-sampling communication:

This is unnecessary because every rank in the DP group already has the full
logits tensor. Communication should only happen after local sampling to gather
compact results.

Cross-DP sampling parallelism:

This was rejected to avoid scheduler, request ownership, and engine-level routing
changes. The proposal is limited to the current DP group.

Environment variables:

This was rejected. vLLM Ascend configuration should use `additional_config` for
this feature.

Automatic fallback from `skip` mode:

This was rejected. `skip` is an explicit performance mode. When incompatible
request parameters are present, the implementation should warn and continue
rather than silently changing execution mode.

### Risks and open questions

- The adapter must keep v1 request lifecycle and sampler state updates in sync.
- Logprobs and speculative decoding have expanded-row and variable-output shape
  contracts that need careful gather/restore logic.
- Skip mode can produce different outputs when users enable skipped parameters;
  documentation and warnings must be clear.
- Fused mode may need staged rollout if Triton-Ascend coverage differs across
  hardware or CANN versions.
- Async scheduling, PP broadcast, and existing Ascend-specific paths such as
  `lmhead_tp_enable()` need explicit validation.

### Feedback Period.

At least one week from 2026-05-18 to 2026-05-25.

### CC List.

Suggested reviewers and area owners:

- `@realliujiaxu` and `@whx-sjtu` for `vllm_ascend/sample`
- `@MengqingCao` for `vllm_ascend/worker`
- `@wangxiyuan` for `vllm_ascend/spec_decode`, config, and overall Ascend architecture

### Any Other Things.

If maintainers prefer a feature branch for this multi-PR effort, this RFC requests a branch such as `rfc/sampling_optimization` after the RFC receives the required maintainer approvals.

Implementation PRs should be split by phase so reviewers can validate the adapter, batch-parallel shape contracts, logprobs support, speculative decoding support, and fused kernel independently.

**Comment by freyfwt** [comment]
CC for review/feedback @realliujiaxu @whx-sjtu @MengqingCao @wangxiyuan 
Please help review the sampler worker spec-decode and config directions when you have bandwidth.
